### ISSUE TITLE
Forge API on Ruby 1.8.7

### ISSUE THREAD

**Issue by Ginja** [puppetlabs/r10k#212]
While creating a git hook to enforce certain versions of modules when users are doing a `git push`, I noticed that r10k was having issues when trying to determine the latest version of a module that was specified as :latest in my control Puppetfile. I tracked down the problem, and it lies with what the forge API returns. I checked this by enabling Faraday logging, and outputting the response.body from the versions method (in lib/r10k/module_repository/forge.rb):

Ruby 2.0.0, works

```
I, [2014-10-29T02:09:07.940478 #4880]  INFO -- : get https://forge.puppetlabs.com/api/v1/releases.json?module=flakrat%2Frepo_dell
D, [2014-10-29T02:09:07.940629 #4880] DEBUG -- request: User-Agent: "Faraday v0.8.9"
I, [2014-10-29T02:09:10.798093 #4880]  INFO -- Status: 200
D, [2014-10-29T02:09:10.798296 #4880] DEBUG -- response: server: "nginx"
date: "Wed, 29 Oct 2014 09:09:10 GMT"
content-type: "application/json;charset=utf-8"
transfer-encoding: "chunked"
connection: "close"
status: "200 OK"
cache-control: "private, must-revalidate, max-age=0"
x-node: "forge-api04-prod"
x-revision: "8c61296"
x-content-type-options: "nosniff"
vary: "Accept-Encoding"
x-forwarded-by: "forge-weblb01-prod.puppetlabs.com"

{"flakrat/repo_dell"=>[{"file"=>"/system/releases/f/flakrat/flakrat-repo_dell-0.1.0.tar.gz", "version"=>"0.1.0", "dependencies"=>[]}, {"file"=>"/system/releases/f/flakrat/flakrat-repo_dell-0.1.1.tar.gz", "version"=>"0.1.1", "dependencies"=>[]}]}
```

Ruby 1.8.7, fails

```
I, [2014-10-29T02:08:18.441347 #7033]  INFO -- : get https://forge.puppetlabs.com/api/v1/releases.json?module=flakrat%2Frepo_dell
D, [2014-10-29T02:08:18.441445 #7033] DEBUG -- request: User-Agent: "Faraday v0.8.9"
I, [2014-10-29T02:08:18.814788 #7033]  INFO -- Status: 200
D, [2014-10-29T02:08:18.814904 #7033] DEBUG -- response: date: "Wed, 29 Oct 2014 09:08:18 GMT"
connection: "close"
content-length: "233"
x-node: "forge-api02-prod"
x-revision: "8c61296"
vary: "Accept-Encoding"
x-forwarded-by: "forge-weblb01-prod.puppetlabs.com"
server: "nginx"
cache-control: "private, must-revalidate, max-age=0"
x-content-type-options: "nosniff"
content-type: "application/json;charset=utf-8"
status: "200 OK"

{"flakrat/repo_dell":[{"file":"/system/releases/f/flakrat/flakrat-repo_dell-0.1.0.tar.gz","version":"0.1.0","dependencies":[]},{"file":"/system/releases/f/flakrat/flakrat-repo_dell-0.1.1.tar.gz","version":"0.1.1","dependencies":[]}]}
```

I have patch for this, which simply checks what was returned and if it's JSON, it will parse it before it attempts to get the version. Would this be something you're interested in me submitting a pull request for? Another workaround is to install a newer version of Ruby using something like RVM, which is what I did.

``` Ruby
# lib/r10k/module_repository/forge.rb
  def versions(module_name)
    response = @conn.get("/api/v1/releases.json", {'module' => module_name})

    response_value = response.body
    # Parse the response if it comes back in JSON format
    response_value = JSON.parse(response.body) unless response.body.class == Hash

    response_value[module_name].map do |version_info|
      version_info['version']
    end
  end
```

**Comment by adrienthebo** [comment]
I'm not clear on what's actually going on - is this an issue where the forge is outputting JSON that Ruby 1.8.7 cannot parse? Could you provide an example Puppetfile that fails on Ruby 1.8.7?

With respect to fixing this, I need to update r10k to use the v3 API anyways; that may solve the issue. Before making any decisions about what changes to apply I would like to nail down a bit more of what's failing.

**Comment by adrienthebo** [comment]
WAIT WRONG BUTTON GAH

**Comment by Ginja** [comment]
AFAICT, it looks like running r10k on a 1.8.7 will return a JSON response when attempting to find a module's version(s), while it looks like the code evaluating the response is expecting a hash.

I'm able to reproduce this problem using two methods:

Method: `r10k puppetfile install`
1. `mkdir -p ~/r10k/modules && touch ~/r10k/Puppetfile`
2.  Specify the following in the Puppetfile:

```
mod "flakrat/repo_dell", :latest
mod "flakrat/repo_epel", :latest 
```

```
r10k puppetfile install
# Result
[R10K::TaskRunner - ERROR] Task #<R10K::Task::Module::Sync:0x7fdbdf4c9d80> failed while running: Invalid version string ''!
```

It's trying to verify the version on a nil object.

Method: `irb`

```
/usr/bin/irb
irb(main):001:0> require 'rubygems'
irb(main):002:0> require 'r10k/puppetfile'
irb(main):003:0> forge = R10K::ModuleRepository::Forge.new
irb(main):004:0> forge.versions('adrien/boolean')
=> [nil]
```

The reason it's returning nil is because the following code won't parse a JSON object:

``` ruby
    response.body[module_name].map do |version_info|
      version_info['version']
    end
```

Everything else seems to work (i.e. version specified, version not specified). This problem only seems to crop up when you specify :latest.

**Comment by adrienthebo** [comment]
That might be an issue with the Faraday middleware, I'll look into this
tomorrow. Thanks for reporting this!

**Comment by Ginja** [comment]
No problem. If it helps, I've tried the following:
1. Downgrading Faraday to 0.8.9 on my Ruby 2.0.0 machine (1.8.7 uses that version)
2. Set the Accept header to 'application/json' with Faraday
3. Tried specifying the transfer-encoding because I noticed it was set to 'chunked' in the response that works, but I'm not sure Faraday supports setting this.

My hunch is that the problem lies with the net/http libraries between the two different versions.

Let me know if you need anything else.

Thanks!
Riley

**Comment by adrienthebo** [comment]
This looks like it's a bug somewhere down inside of MultiJson; the Faraday middleware is trying to parse the response using MultiJson, swallowing the error, and returning the body as-is:

```
     9: def parse(body)
    10:   require 'pry'
 => 11:   binding.pry
    12:   ::MultiJson.load(body, @options) rescue body
    13: end

[1] pry(#<FaradayMiddleware::MultiJson::ParseJson>)> ::MultiJson.load(body, @options)
ArgumentError: Did not recognize your adapter specification.
from /home/adrien/development/r10k/.bundle/ruby/1.8/gems/multi_json-1.8.4/lib/multi_json.rb:108:in `load_adapter'

```

I'll keep digging.

**Comment by adrienthebo** [comment]
Per 2cfe6780ffd8b72c400a03933d9a7a695a57160a, I brought this on myself: 

```
[spec] Exclude VCR tests on Ruby 1.8.7

For some reason VCR isn't working in the given setup. This is a specific
interaction with VCR so I'm going to sweep it under the rug. I will
certainly regret this later.
```

This is happening because Ruby 1.8.7 doesn't ship with the JSON gem, whereas 1.9 ships with JSON in core. r10k requires json_pure to ensure there's always a JSON library available, but it looks like json_pure confuses Faraday into thinking that the JSON gem is available, tries to load it, and then explodes. In turn, the multi_json Faraday adapter tries to parse the provided JSON and rescues all errors and returns the string, thus squelching the error message and returning useless information.

I think that forcing the use of json_pure on 1.8 is the easiest route to fix this; r10k doesn't parse much JSON, what JSON it does parse is pretty small, and it's not even close to being a performance bottleneck.

**Comment by Ginja** [comment]
Nice sleuthing!

**Comment by adrienthebo** [comment]
Pull request filed as GH-218.

**Comment by adrienthebo** [comment]
Merged into 1.3.x in 42d5a9b; this will be released in 1.3.5 if I do another 1.3.x, otherwise it'll be released in 1.4.0. Thanks for reporting this!

**Comment by adrienthebo** [comment]
This has been released in 1.3.5.
